fdbserver/DataDistribution.actor.cpp

```diff
@@ -641,36 +645,122 @@ struct DataDistributor : NonCopyable, ReferenceCounted<DataDistributor> {
+}
+
+std::unordered_map<UID, std::shared_ptr<DDAudit>> getAuditsForType(Reference<DataDistributor> self,
+                                                                   AuditType auditType) {
+	ASSERT(self->audits.contains(auditType));
+	return self->audits[auditType];
+}
+
+void runAuditStorage(Reference<DataDistributor> self,
+                     AuditStorageState auditStates,
+                     int retryCount,
+                     std::string context);
+ACTOR Future<Void> waitForAuditStorage(Reference<DataDistributor> self, UID auditID, AuditType auditType);
+ACTOR Future<Void> auditStorageCore(Reference<DataDistributor> self,
+                                    UID auditID,
+                                    AuditType auditType,
+                                    std::string context,
+                                    int currentRetryCount);
+ACTOR Future<UID> launchAudit(Reference<DataDistributor> self, KeyRange auditRange, AuditType auditType);
 ACTOR Future<Void> auditStorage(Reference<DataDistributor> self, TriggerAuditRequest req);
-ACTOR Future<Void> loadAndDispatchAuditRange(Reference<DataDistributor> self,
-                                             std::shared_ptr<DDAudit> audit,
-                                             KeyRange range);
-ACTOR Future<Void> scheduleAuditForRange(Reference<DataDistributor> self,
-                                         std::shared_ptr<DDAudit> audit,
-                                         KeyRange range);
+ACTOR Future<Void> loadAndDispatchAudit(Reference<DataDistributor> self,
+                                        std::shared_ptr<DDAudit> audit,
+                                        KeyRange range);
+ACTOR Future<Void> partitionAuditJobByServer(Reference<DataDistributor> self, std::shared_ptr<DDAudit> audit);
+ACTOR Future<Void> scheduleAuditPerServer(Reference<DataDistributor> self, std::shared_ptr<DDAudit> audit);
+ACTOR Future<Void> partitionAuditJobByRange(Reference<DataDistributor> self,
+                                            std::shared_ptr<DDAudit> audit,
+                                            KeyRange range);
+ACTOR Future<Void> scheduleAuditOnRange(Reference<DataDistributor> self,
+                                        std::shared_ptr<DDAudit> audit,
+                                        KeyRange range);
 ACTOR Future<Void> doAuditOnStorageServer(Reference<DataDistributor> self,
                                           std::shared_ptr<DDAudit> audit,
                                           StorageServerInterface ssi,
                                           AuditStorageRequest req);
 
 ACTOR Future<Void> resumeStorageAudits(Reference<DataDistributor> self) {
-	state std::vector<Future<Void>> fs;
+	if (self->initData->auditStates.empty()) {
+		self->auditInitialized.send(Void());
+		return Void();
+	}
+	cancelAllAuditsInAuditMap(self); // cancel existing audits
```

Comment:
Should the map be empty as this moment, since we are initializing it?

Reply:
Not necessary. Existing self->audits is dropped and a new self->audits is available i.i.f. a new DataDistributor is created.  `cancelAllAuditsInAuditMap` is called i.i.f. resumeStorageAudits. However, resumeStorageAudits is not necessarily called when a new DataDistributor is established.

Reply:
`resumStorageAudits` is called only during DD initialization if I didn't miss other places, in addition, this would counter the above `ASSERT(!self->auditInitialized.getFuture().isReady());` if it is not called during initialization.

Reply:
Yes. Changed.